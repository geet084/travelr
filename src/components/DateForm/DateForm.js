import React, { Component } from 'react';
import PropTypes from 'prop-types';

export class DateForm extends Component {
  constructor() {
    super();
    this.monthRef = React.createRef();
    this.dayRef = React.createRef();
    this.yearRef = React.createRef();
    this.state = {
      month: '',
      day: '',
      year: '',
    };
  }

  componentWillUpdate(prevProps, prevState) {
    const { month, day, year } = prevState;
    const isFullDate = month.length === 2 && day.length === 2 && year.length === 4;

    if (isFullDate) this.props.updateUserDate(`${month}-${day}-${year}`);
  }

  buildDateInputs(dateTypes) {
    return Object.keys(dateTypes).map(key => {
      let dateType = dateTypes[key];
      return (
        <input
          id={key}
          key={key}
          max={dateType.max}
          maxLength={dateType.length}
          min={dateType.min}
          onBlur={this.handleDateInput}
          onChange={this.handleDateInput}
          placeholder={dateType.placeholder}
          ref={`${key}Ref`}
          type='text'
          value={this.state[key]}
        />
      );
    });
  }

  handleDateInput = ({ target }) => {
    const value = this.handleNumberVerification(target);
    
    this.handleInputFocus(target, value);
    this.setState({ [target.id]: value });
  };

  handleNumberVerification = ({ value, min, max }) => {
    value = value.trim()
    const isNumber = Number(value) >= 0 && Number(value) <= Number(max);
    const higherThanMax = Number(value) >= Number(max);
    const lessThanMin = value.length === max.length && Number(value) < Number(min);
    if (lessThanMin) return min
    else if (higherThanMax) return max
    else if(!isNumber) return NaN
    else return value
  }

  handleInputFocus(target, value) {
    const isMaxLength = value.length === target.maxLength

    if (target.id === 'month' && isMaxLength) this.refs.dayRef.focus();
    if (target.id === 'day' && isMaxLength) this.refs.yearRef.focus();
    if (target.id === 'year' && isMaxLength) this.refs.yearRef.blur();
  }

  render() {
    const dateTypes = {
      month: { placeholder: 'mm', length: 2, min: '01', max: '12' },
      day: { placeholder: 'dd', length: 2, min: '01', max: '31' },
      year: { placeholder: 'yyyy', length: 4, min: '0100', max: '2999' }
    };
    const dateInputs = this.buildDateInputs(dateTypes);

    return (
      <div className='date-form'>
        <form id='form'>
          {dateInputs}
        </form>
      </div>
    );
  }
}

DateForm.propTypes = {
  updateUserDate: PropTypes.func
};

export default DateForm;