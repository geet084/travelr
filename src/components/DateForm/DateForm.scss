.date-form {
  background-color: #fff;
  width: 170px;
  position: relative;
  top: -25px;
  right: -217px;
  margin: 0;
}

#form {
  border-bottom: 1px solid rgba(0, 0, 0, .2);
  width: 130px;
  padding-bottom: 11px;
  padding-left: 10px;
}
input {
  border: none;
  width: 35px;
  outline: none;
  font-size: 15px;
  padding: none;
  margin: none;
  text-align: left;
}

input::placeholder {
  opacity: .5;
  font-weight: 300;
  color: red;
}
#day {
  width: 20px;
}
#month {
  width: 35px;
}
span {
  color: #fff;
}
.first-span {
  margin-left: -10px;
}
