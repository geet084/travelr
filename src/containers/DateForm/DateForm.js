import React, { Component } from 'react'
import { connect } from 'react-redux';
import { setUserInfo } from '../../actions'
import DateInput from 'date-input';

export class DateForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      minDate: '0000-01-01',
      maxDate: '',
      userDate: this.props.userDate,
      elapsedDays: this.props.elapsedDays,
    }
  }
  
  componentDidMount = () => {
    this.setMaxDate();
  }

  setMaxDate = () => {
    const date = new Date(this.props.today);
    let month = '' + (date.getMonth() + 1);
    let day = '' + (parseInt(date.getDate()) + 1);
    const year = date.getFullYear();

    if (month.length < 2) month = '0' + month;
    if (day.length < 2) day = '0' + day;
    const maxDate = [year, month, day].join('-');

    this.setState({ maxDate });
  }

  handleDate = (e) => {
    let userDate = new Date(e);
    userDate.setDate(userDate.getDate());
    const singleDay = (1000 * 60 * 60 * 24);
    const thisTime = new Date();
    const diff = thisTime.getTime() - userDate.getTime();
    let elapsedDays = (diff / singleDay);
    const date = e.split('');

    elapsedDays = this.fixElapsedDays(elapsedDays, date);
    userDate = this.fixDateError(userDate);

    this.setState({ userDate, elapsedDays });
    this.props.setUserInfo({ userDate, elapsedDays })
  }

  fixElapsedDays = (elapsedDays, date) => {
    let days;
    if (date.length < 10 || elapsedDays <= 0) days = 0;
    else days = elapsedDays.toFixed(2);
    return days;
  }

  fixDateError = (date) => {
    let days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
    date = date.toDateString().split(' ');
    date[0] = (days[days.indexOf(date[0]) + 1]);
    date[2] = (parseInt(date[2]) + 1).toString();

    return date.join(' ');
  }

  render() {
    const { elapsedDays, maxDate, minDate, userDate } = this.state;
    const days = Math.floor(elapsedDays)
    const hours = Math.floor(((elapsedDays - days) * 24) - 7)
    const total = ((days + (hours / 24)) * 24000);

    return (
      <div className="date-form">
        <div className="counter">
          <DateInput shouldValidate minDate={minDate} maxDate={maxDate} onChange={this.handleDate} />
        </div>
        {
          elapsedDays !== 0 &&
          <div>
            <p className="test">Time elapsed since {userDate}: </p>
            <p className="test">
              <span>{days.toLocaleString()} days and </span>
              <span> ~{hours.toLocaleString()} hours</span>
            </p>
            <p className="test">Approx. distance you've gone on Earth: </p>
            <p className="test"> {total.toLocaleString()} miles</p>
          </div>
        }
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  setUserInfo: (userInfo) => dispatch(setUserInfo(userInfo)),
})

export const mapStateToProps = (state) => ({
  userDate: state.userInfo.userDate,
  elapsedDays: state.userInfo.elapsedDays,
})

export default connect(mapStateToProps, mapDispatchToProps)(DateForm);