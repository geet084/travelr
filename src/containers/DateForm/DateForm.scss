
.counter {
  background-color: #fff;
  width: 170px;
  margin: 0 auto;
}

.date-form {
  margin: 25px 100px;
}