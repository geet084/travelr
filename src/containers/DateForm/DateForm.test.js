import React from 'react';
import { shallow } from 'enzyme';
import DateForm from './DateForm';
import { mapStateToProps, mapDispatchToProps } from './DateForm'


describe('DateForm', () => {
  let wrapper;
  const mockToday = 'Mon Oct 01 2018'

  beforeEach(() => {
    wrapper = shallow(<DateForm today={mockToday} />)
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const mockUserInfo = 'I am a user'
    const mappedProps = mapDispatchToProps(mockDispatch);

    mappedProps.setUserInfo(mockUserInfo);
    expect(mockDispatch).toHaveBeenCalled()
  })

  it('should mapStateToProps', () => {
    const mockState = {
      userInfo: {
        userDate: 'Oct, 1, 2018',
        elapsedDays: 199
      },
      extra: 'stuff',
      things: 'things'
    }
    const expected = {
      userDate: 'Oct, 1, 2018',
      elapsedDays: 199,
    }
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected)
  })
})