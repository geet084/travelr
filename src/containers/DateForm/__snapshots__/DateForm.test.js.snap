// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateForm should match snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
